lib/model/message.dart

```diff
@@ -409,22 +409,59 @@ class MessageStoreImpl extends HasChannelStore with MessageStore, _OutboxMessage
     // from the event queue, so there's inherently a race.
     //
     // If the fetched message reflects changes we haven't yet heard from the
-    // event queue, then it doesn't much matter which version we use: we'll
-    // soon get the corresponding events and apply the changes anyway.
-    // But if it lacks changes we've already heard from the event queue, then
-    // we won't hear those events again; the only way to wind up with an
+    // event queue, then normally (see [1]) it doesn't much matter which version
+    // we use: we'll soon get the corresponding events and apply the changes
+    // anyway. But if it lacks changes we've already heard from the event queue,
+    // then we won't hear those events again; the only way to wind up with an
     // updated message is to use the version we have, that already reflects
-    // those events' changes.  So we always stick with the version we have.
+    // those events' changes.  So we always stick with the version we have. [1]
+    //
+    // [1] With one exception: if the version we have was in an unsubscribed
+    //     channel when we got it or sometime since, we take the fetched version
+    //     instead. That's because our version might be stale; we don't expect
+    //     update events for messages in unsubscribed channels.
     for (int i = 0; i < messages.length; i++) {
       final message = messages[i];
-      messages[i] = this.messages.putIfAbsent(message.id, () {
-        message.matchContent = null;
-        message.matchTopic = null;
-        return message;
-      });
+
+      // Whether we had marked the version we have (if any) as potentially
+      // stale; see [1] above.
+      final ourVersionMightBeStale =
+        message is StreamMessage
+        // Side effect: update our "potentially stale" knowledge based on
+        // whether the channel is subscribed right now.
+        && (subscriptions[message.streamId] != null
+            ? _maybeStaleChannelMessages.remove(message.id)
+            : !_maybeStaleChannelMessages.add(message.id));
+
+      if (ourVersionMightBeStale) {
+        this.messages[message.id] = _stripMatchFields(message);
+      } else {
+        messages[i] = this.messages.putIfAbsent(message.id, () {
+          return _stripMatchFields(message);
+        });
+      }
     }
   }
 
+  /// Messages in [messages] whose data stream is or was presumably broken
+  /// by the message being in an unsubscribed channel.
+  ///
+  /// We don't expect update events for messages in unsubscribed channels,
+  /// so if some of these maybe-stale messages appear in a fetch,
+  /// we'll always clobber our stored version with the fetched version.
+  /// See implementation comment in [reconcileMessages].
+  ///
+  /// (We have seen a few such events, actually --
+  /// maybe because the channel was recently subscribed? --
+  /// but not consistently, and we're not supposed to rely on them.)
```

Comment:
```suggestion
  /// (We have seen a few such events, actually --
  /// maybe because the channel was recently unsubscribed? --
  /// but not consistently, and we're not supposed to rely on them.)
```

?